### PR TITLE
Results of parsing _ViewStart files should be cached

### LINKED ISSUE #1016: Results of parsing _ViewStart files should be cached
This might involve taking advantage of `IFileInfoCache`, but the plan would be to not-reparse files if they haven't changed.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #1016

### CODE REVIEW

**Comment by pranavkm** [comment]
Updated.

File: src/Microsoft.AspNet.Mvc.Razor.Host/Directives/DefaultCodeTreeCache.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+using Microsoft.Framework.Cache.Memory;
+
+namespace Microsoft.AspNet.Mvc.Razor.Directives
+{
+    /// <summary>
+    /// Default implementation of <see cref="ICodeTreeCache"/>.
+    /// </summary>
+    public class DefaultCodeTreeCache : ICodeTreeCache
+    {
+        private static readonly MemoryCacheOptions _memoryCacheOptions = new MemoryCacheOptions
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor.Host/Directives/DefaultCodeTreeCache.cs]: `private static readonly` = PascalCase everywhere else, same comment for `_slidingExpirationDuration`

File: src/Microsoft.AspNet.Mvc.Razor.Host/Directives/DefaultCodeTreeCache.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        /// <summary>
+        /// Initializes a new instance of <see cref="DefaultCodeTreeCache"/>.
+        /// </summary>
+        /// <param name="fileProvider">The application's <see cref="IFileProvider"/>.</param>
+        public DefaultCodeTreeCache(IFileProvider fileProvider)
+            : this(fileProvider, _memoryCacheOptions)
+        {
+        }
+
+        internal DefaultCodeTreeCache(IFileProvider fileProvider,
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor.Host/Directives/DefaultCodeTreeCache.cs]: `// Internal for testing`

Note: this is only used in 1 test.

File: src/Microsoft.AspNet.Mvc.Razor.Host/Directives/DefaultCodeTreeCache.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            _fileProvider = fileProvider;
+            _codeTreeCache = new MemoryCache(options);
+        }
+
+        /// <inheritdoc />
+        public CodeTree GetOrAdd([NotNull] string pagePath,
+                                 [NotNull] Func<IFileInfo, CodeTree> getCodeTree)
+        {
+            return _codeTreeCache.GetOrSet(pagePath,
+                                           getCodeTree,
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor.Host/Directives/DefaultCodeTreeCache.cs]: nit: no need to break up lines.

File: src/Microsoft.AspNet.Mvc.Razor.Host/Directives/ICodeTreeCache.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNet.FileProviders;
+using Microsoft.AspNet.Razor.Generator.Compiler;
+
+namespace Microsoft.AspNet.Mvc.Razor.Directives
+{
+    /// <summary>
+    /// A cache for parsed <see cref="CodeTree"/>.
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor.Host/Directives/ICodeTreeCache.cs]: `<see cref="CodeTree"/>s.`

**Comment by NTaylorMullen** [comment]
:shipit:

**Pull request closed without merging**